src/compiler/scala/tools/nsc/backend/jvm/GenBCode.scala

```diff
@@ -3,470 +3,71 @@
+
+  val codeGen = new { val bTypes: GenBCode.this.bTypes.type = GenBCode.this.bTypes } with CodeGen[global.type](global)
+
+  val postProcessor = new { val bTypes: GenBCode.this.bTypes.type = GenBCode.this.bTypes } with PostProcessor
 
   val phaseName = "jvm"
 
   override def newPhase(prev: Phase) = new BCodePhase(prev)
 
-  final class PlainClassBuilder(cunit: CompilationUnit) extends SyncAndTryBuilder(cunit)
-
   class BCodePhase(prev: Phase) extends StdPhase(prev) {
-
-    override def name = phaseName
     override def description = "Generate bytecode from ASTs using the ASM library"
-    override def erasedTypes = true
-
-    private var bytecodeWriter  : BytecodeWriter   = null
-    private var mirrorCodeGen   : JMirrorBuilder   = null
-    private var beanInfoCodeGen : JBeanInfoBuilder = null
 
-    /* ---------------- q1 ---------------- */
-
-    case class Item1(arrivalPos: Int, cd: ClassDef, cunit: CompilationUnit) {
-      def isPoison = { arrivalPos == Int.MaxValue }
-    }
-    private val poison1 = Item1(Int.MaxValue, null, null)
-    private val q1 = new java.util.LinkedList[Item1]
+    override val erasedTypes = true
 
-    /* ---------------- q2 ---------------- */
+    def apply(unit: CompilationUnit): Unit = codeGen.genUnit(unit)
 
-    case class Item2(arrivalPos:     Int,
-                     mirror:         asm.tree.ClassNode,
-                     plain:          asm.tree.ClassNode,
-                     bean:           asm.tree.ClassNode,
-                     sourceFilePath: String,
-                     outFolder:      scala.tools.nsc.io.AbstractFile) {
-      def isPoison = { arrivalPos == Int.MaxValue }
-    }
-
-    private val poison2 = Item2(Int.MaxValue, null, null, null, null, null)
-    private val q2 = new _root_.java.util.LinkedList[Item2]
+    override def run(): Unit = {
+      val bcodeStart = Statistics.startTimer(BackendStats.bcodeTimer)
 
-    /* ---------------- q3 ---------------- */
+      initialize()
 
-    /*
-     *  An item of queue-3 (the last queue before serializing to disk) contains three of these
-     *  (one for each of mirror, plain, and bean classes).
-     *
-     *  @param jclassName  internal name of the class
-     *  @param jclassBytes bytecode emitted for the class SubItem3 represents
-     */
-    case class SubItem3(
-      jclassName:  String,
-      jclassBytes: Array[Byte]
-    )
+      val genStart = Statistics.startTimer(BackendStats.bcodeGenStat)
+      super.run() // invokes `apply` for each compilation unit
+      Statistics.stopTimer(BackendStats.bcodeGenStat, genStart)
 
-    case class Item3(arrivalPos: Int,
-                     mirror:     SubItem3,
-                     plain:      SubItem3,
-                     bean:       SubItem3,
-                     outFolder:  scala.tools.nsc.io.AbstractFile) {
+      postProcessor.postProcessAndSendToDisk()
```

Comment:
That's a good suggestion. We can also run local optimizations in this mode. Can we do that in a follow-up PR?